src/cargo/core/compiler/build_config.rs

```diff
@@ -73,10 +73,12 @@ impl BuildConfig {
                  its environment, ignoring the `-j` parameter",
             )?;
         }
-        let jobs = jobs.or(cfg.jobs).unwrap_or(::num_cpus::get() as u32);
-        if jobs == 0 {
-            anyhow::bail!("jobs may not be 0");
-        }
+        let jobs = match jobs.or(cfg.jobs) {
+            None => ::num_cpus::get() as u32,
+            Some(0) => anyhow::bail!("jobs may not be 0"),
+            Some(j) if j < 0 => (::num_cpus::get() as i32 + j).max(1) as u32,
```

Comment:
On overflow do we want to max out on the number of CPUs instead of 1?
```
Some(j) if j < 0 => (::num_cpus::get() as i32 + j).max(::num_cpus::get()) as u32,
```

Reply:
If someone does `-j -8`, expecting to run on a 16+ core system and the command invocation runs on a 2 core system, like most free CIs, then I would expect it to "pick the smallest valid value" or error rather than "pick highest valid value".  That seems most counter to the intent.

Reply:
Makes sense.